File: framework/utils/WaitUtils.ts
import { Page } from '@playwright/test';

export class WaitUtils {
  /**
   * Pause execution for a set number of milliseconds
   */
  public static async sleep(ms: number): Promise<void> {
    return new Promise(resolve => setTimeout(resolve, ms));
  }

  /**
   * Poll a custom condition function until it returns true or times out
   */
  public static async waitForCondition(
    condition: () => Promise<boolean> | boolean,
    timeout = 10000,
    pollInterval = 500
  ): Promise<boolean> {
    const startTime = Date.now();
    while (Date.now() - startTime < timeout) {
      if (await condition()) {
        return true;
      }
      await this.sleep(pollInterval);
    }
    throw new Error(`Condition was not met within the timeout of ${timeout}ms.`);
  }

  /**
   * Wait for network idle state (no active network requests for at least 500ms)
   */
  public static async waitForNetworkIdle(page: Page, timeout = 10000): Promise<void> {
    console.log(`[WaitUtils] Waiting for network idle...`);
    await page.waitForLoadState('networkidle', { timeout });
  }

  /**
   * Wait until the number of elements matching the selector equals the target count
   */
  public static async waitForElementCount(
    page: Page,
    selector: string,
    expectedCount: number,
    timeout = 10000
  ): Promise<void> {
    console.log(`[WaitUtils] Waiting for element count of "${selector}" to be ${expectedCount}`);
    await this.waitForCondition(async () => {
      const count = await page.locator(selector).count();
      return count === expectedCount;
    }, timeout);
  }
}

File: framework/utils/wait_utils.py
from __future__ import annotations

import time
from collections.abc import Callable

from playwright.sync_api import Page


class WaitUtils:
    @staticmethod
    def sleep(milliseconds: float) -> None:
        time.sleep(milliseconds / 1000)

    @staticmethod
    def wait_for_condition(
        condition: Callable[[], bool],
        timeout: float = 10_000,
        poll_interval: float = 500,
    ) -> bool:
        deadline = time.monotonic() + timeout / 1000
        while time.monotonic() < deadline:
            if condition():
                return True
            WaitUtils.sleep(poll_interval)
        raise TimeoutError(f"Condition was not met within {timeout}ms")

    @staticmethod
    def wait_for_network_idle(page: Page, timeout: float = 10_000) -> None:
        page.wait_for_load_state("networkidle", timeout=timeout)

    @staticmethod
    def wait_for_element_count(
        page: Page, selector: str, expected_count: int, timeout: float = 10_000
    ) -> None:
        WaitUtils.wait_for_condition(
            lambda: page.locator(selector).count() == expected_count,
            timeout,
        )
